Gaussian processes

bwengals opened the pull request:
Adding a basic set of covariance functions / kernels for doing Gaussian Process regression.  They follow a similar API as GPy and GPFlow.  Some main differences are:
  - Covariance function objects allow for multiplication and addition with a scalar (or PyMC3 variable) or a numpy array of the right size.
  - Covariance functions aren't named, their parameters are PyMC3 variables which have names.

At this point, Gibbs and Linear aren't correct yet.  Hope to eventually have a fairly complete suite of covariance functions.

This WIP PR also includes some notebook examples showing how to use them, and an example of fitting a GP.  

Please, feedback and suggestions welcome!

bwengals committed:
initial gp commit

- MvNormal2,multivariate normal using cholesky, solve_triangular instead of direct inversion
- covariance functions with add, product methods that work on scalars and other covariance objects
- basic example / test of new functionality

bwengals committed:
initial gp commit

- MvNormal2,multivariate normal using cholesky, solve_triangular instead of direct inversion
- covariance functions with add, product methods that work on scalars and other covariance objects
- basic example / test of new functionality

bwengals committed:
moved MvNormal2 class to GP

bwengals committed:
moved MvNormal2 class to GP

bwengals committed:
fixed example

bwengals committed:
fixed example

bwengals committed:
minor bug fix -- need to add 'vars' argument to call to 'guess_scaling'

bwengals committed:
minor bug fix -- need to add 'vars' argument to call to 'guess_scaling'

bwengals committed:
removed MvNormal2 from distributions library

bwengals committed:
removed MvNormal2 from distributions library

bwengals committed:
import fix for GP class

bwengals committed:
import fix for GP class

bwengals committed:
moved example into notebook, added notebook with model with latent variable explicit

bwengals committed:
moved example into notebook, added notebook with model with latent variable explicit

bwengals committed:
latent example working

bwengals committed:
latent example working

bwengals committed:
added some more stationary cov functions

bwengals committed:
added some more stationary cov functions

bwengals committed:
add notebook examples

bwengals committed:
add notebook examples

bwengals committed:
added cosine covariance function

bwengals committed:
added cosine covariance function

bwengals committed:
added linear covariance function

bwengals committed:
added linear covariance function

bwengals committed:
updated covariance examples to show lin and cos cov functs

bwengals committed:
updated covariance examples to show lin and cos cov functs

bwengals committed:
added input warping covariance

bwengals committed:
added input warping covariance

bwengals committed:
updated cov example w warp

bwengals committed:
updated cov example w warp

bwengals committed:
allow warping function to take pymc3 variables as parameters

bwengals committed:
allow warping function to take pymc3 variables as parameters

bwengals committed:
very basic basis function kernel

bwengals committed:
very basic basis function kernel

bwengals committed:
cov examples update

bwengals committed:
cov examples update

bwengals committed:
remove warning message in __init__ about using a different branch

bwengals committed:
basisfunctioncov just for basic case, added message and cleaned up

bwengals committed:
fixed init file

bwengals committed:
improve cov examples

bwengals committed:
added prediction to gp example

bwengals committed:
removed gp class that used cholesky, more work on covariance functions

bwengals committed:
some cleanup, no major changes

bwengals committed:
improved example, using product of Matern52 and Cosine

bwengals committed:
removed small fix to hmc.py, removed theano.slinalg import in gp branhces multivariate.py

twiecki commented on pymc3/__init__.py:
```diff
@@ -23,6 +23,7 @@
 from .tests import test
 
 from . import glm
+from . import gp
```
Can we move this to `pymc3.models`?

bwengals replied:
Absolutely.  I wasn't sure where things should go

fonnesbeck replied:
I think `gp` is fine. I don't think we want both `pymc3.model` and `pymc3.models`.

twiecki commented:
I've only skimmed so far but this is very exciting. Really like the compositionality of kernels. Some high-level points:
- [x] Should move to `pymc3.models.gp`.
- [ ] The examples should live in `docs/sources/notebooks` and we should add a separate category in `docs/sources/examples.rst` for them.
- [x] Should we renamed `covariance` to `kernel`? I feel like the latter is more commonly used in this context and has less risk of being confused with covariance used elsewhere. Then again, it directly links to the covariance being passed to `MvNormal` so maybe it's actually more consistent that way.
- [ ] How does NUTS perform?
- [ ] Needs unit-tests.
- [x] Rebase to `3.1` branch, soon to be merged into master.

twiecki commented on pymc3/gp/cov.py:
```diff
@@ -0,0 +1,222 @@
+        return tt.sqrt(r2 + 1e-12)
+
+
+class ExpQuad(Stationary):
```
Doc-strings, can we copy & paste those from `GPy`?

bwengals replied:
These are based on [GPFlow](https://github.com/GPflow/GPflow/blob/master/GPflow/kernels.py).  It was easy to adapt their tensorflow to theano.  For most of them it should be a copy/paste job.

bwengals commented:
- Sounds good
- Sounds good
- This I was on the fence with too, for the exact reasons you mentioned.  They are usually called kernels, but they literally are covariances.  I'm okay with either naming convention though.
- Not really sure yet.  I found NUTS would hang or get stuck sometimes.  Slice seems to work the best.  This is just a first impression though.
- Agree
- Sounds good

Awesome!

twiecki commented:
OK, NUTS might need some help tuning, we can leave that for now, it's a more general problem we're working on.

fonnesbeck commented:
OK, I see on 3.1 we now have a `models` submodule. I will provide feedback there; I don't think that's a good idea.

twiecki commented:
@fonnesbeck The `models` submodule or placing GP there? I actually have my concerns on the former as well. 

Maybe we can make progress this way too: Thinking about this PR very narrowly, it does not implement GPs, but rather functionality to construct covariance matrices with certain structure. These just happen to be used in GPs. @bwengals thoughts?

fonnesbeck commented:
I think GPs are unique enough (in the sense that they are models of functions) to have their own submodule. I'd be open to `pymc3.np`, or similar, thinking ahead to when we might have Dirichlet processes, etc.

twiecki commented:
Definitely, just this isn't GP, it's covariance matrix generators. Just thinking of how to name / structure this. I'm also not sure Dirichlet Process would be a submodule, it would just be a another distribution. And we already have an example of a DPGMM. Since we're not providing models but building blocks for models, the naming is a bit different.

fonnesbeck commented:
`kernels` might work. I would favor something along the lines of `pymc3.gp` alongside `pymc3.kernels`.

bwengals commented:
I'm thinking that including a `gp` submodule where the kernels/covariances would be appropriate if there was some convenience function modeled after `glm`.  Perhaps something like:
```
with pm.Model() as model2:
    l = pm.HalfNormal('l')
    cov_f = pm.gp.Matern52(1,l)
    gp1 = pm.gp.GP('gp1', mean=0, latent_cov=cov_f, likelihood='poisson', data=X, observed=y)
```
for a Poisson model, or 
```
with pm.Model() as model2:
    ...
    gp2 = pm.gp.GP('gp2', mean=mean_func, latent_cov=cov_f, noise_cov=cov_n, likelihood='normal', observed=y, data=X, predict=Z)
```
for a normal model.  I started this functionality, but got a bit bogged down in the details and decided to hold off for the time being.

If not a `gp` directory, maybe use a `kernels` or `cov_funcs` directory in the meantime?  It's also possible that as more kernels are added, cov.py will need to be broken into multiple files.

bwengals committed:
this example broken

twiecki commented:
I merged 3.1 into master, so you can just rebase off of master again.

I think such functionality would be great. I'm not opposed to a gp subdirectory, just exploring ideas. If that gets added it sounds right to have a gp submodule. We can leave that for now.

twiecki commented:
Also, the likelihood should probably be a dist:
```python
with pm.Model() as model2:
    l = pm.HalfNormal('l')
    cov_f = pm.gp.Matern52(1,l)
    gp1 = pm.gp.GP('gp1', mean=0, latent_cov=cov_f, likelihood=pm.Poisson, data=X, observed=y)
```

twiecki commented:
@bwengals I think the only things left then are unittests and doc-stirngs.

bwengals commented:
Hi @twiecki, haven't had time to work on this lately, sorry about that!  I should be able to finish these two things up within the next couple days.   

I have a question though.  How do you guys want to handle additions to the GP library after the covariance functions get merged into master?  Should new PRs be opened for each additional feature?  Or should development continue on a GP branch?

twiecki commented:
@bwengals Sounds good. Regarding PRs: it's really up to you. If you're not super sure about the design and want to experiment perhaps a separate branch is the way to go.

fonnesbeck commented:
I would merge new functionality into master as they are ready for use, and keep work-in-progress on a separate branch. We might as well get new features in the hands of users on the master branch if they are working so that they can be more thoroughly tested. Thanks for your work on this.

bwengals commented:
Looking forward to getting this in.  Thanks guys, that sounds good.

bwengals committed:
fit linear

bwengals committed:
removed gibbs, fixedbasis. these need more work

bwengals committed:
doc strings finished, documentation compiles correctly

twiecki commented:
@bwengals Let us know when you want us to take another look.

bwengals commented:
will do

bwengals committed:
added polynomial kernel, moved covariance notebook to docs/source/notebooks

bwengals committed:
added cov func notebook to examples.rst

bwengals committed:
finished GP and cov tutorial

bwengals committed:
small fixes to notebooks

bwengals committed:
finished introduction, moved GP smoothing example into GP section in docs

bwengals commented:
@twiecki @fonnesbeck I think things are at a good stopping point for the time being.  Documentation looks good to me at least.  Unit tests are outstanding, but the covariance notebook sort of serves this purpose.  Please, let me know if you'd like any changes, particularly to the examples.  I'm not sure if everything I'm doing in there is the "best" PyMC3 way to do it. 

Oh, what was the decision on structure?  ```pm.gp``` next to ```pm.kernels```? Or ```pm.gp.cov```?  Also, kernel vs. covariance function?

bwengals committed:
Merge branch 'master' into gaussian-processes

twiecki commented on docs/source/examples.rst:
```diff
@@ -37,6 +36,14 @@ GLM
+.. toctree::
+   notebooks/GP-introduction.ipynb
+   notebooks/GP-covariances.ipynb
+   notebooks/GP-smoothing.ipynb
```
I don't think this NB is added.

twiecki commented:
@bwengals Looks good. lets stick with `pm.gp` and `covariance`, as you have it now.
Note that the permissions of `multivariate.py` need to be still be fixed (reverted) and a NB seems to miss.

twiecki commented:
Also need to fix headings in GP-introduction.ipynb. First, it needs a single header at the top (with `#`), then all following headings should be sub-headings (i.e. `##` or more).

bwengals committed:
GLM-hierarchicical-advi-minibatch had +x permissions, chmod -x

bwengals committed:
fixed markdown headings in GP-introduction.ipynb

bwengals committed:
added polynomial cov to api.rst

bwengals committed:
added polynomial to __all__ in cov.py

bwengals committed:
Merge branch 'master' into gaussian-processes

bwengals committed:
reordered covariances in api.rst

bwengals commented:
Sorry about that!  I *think* the issues you mentioned are resolved now, after a merging ```master``` into ```gaussian-processes```.

twiecki commented:
```
+++[[ true == \t\r\u\e ]]
+++. ./scripts/lint.sh
++++pylint --rcfile=.pylintrc pymc3
************* Module pymc3.gp.cov
pymc3/gp/cov.py:2: [W0611(unused-import), ] Unused import copy
pymc3/gp/cov.py:4: [W0611(unused-import), ] Unused import theano****
```

twiecki commented:
Also, now two files changed permissions from before. they should be 755.

bwengals commented:
Which files?  How/where are you seeing the file permission issue?  Thanks for your patience while I'm being a noob.

twiecki commented:
On github, if you switch to the file view of your PR. multivariate.py

On Jan 27, 2017 12:32 AM, "Bill Engels" <notifications@github.com> wrote:

> Which files? How/where are you seeing the file permission issue? Thanks
> for your patience while I'm being a noob.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc3/pull/1674#issuecomment-275547484>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AApJmEZxW5VI5hUZUPog2XryHQliM6RPks5rWS0YgaJpZM4Lldk9>
> .
>

bwengals committed:
merge fix to pymc3/__init__.py

bwengals committed:
set permission 755 on multivariate.py

bwengals committed:
removed unused imports

bwengals committed:
added unittests

bwengals committed:
usage is now e.g.: pm.gp.cov.ExpQuad

bwengals committed:
added assertion checking dim and X.shape in _slice

bwengals committed:
removed problematic test

bwengals committed:
maybe fixed issue w python3 build?

bwengals committed:
removed dimension checks.  dim checking inputs not obvious with warpedinput cov.  tests passing, looks good

bwengals commented:
Ok, fixed the issue with ```multivariate.py```.  Thanks for being patient.  Over the last few commits, I added unit tests, had some issues with getting them to pass in the python3.6 build, then got that fixed.  As far as I can tell things look good.

twiecki commented:
@bwengals Many thanks, this is a great addition!

bwengals commented:
@twiecki No thank you!  I look forward to contributing more.
